private/buf/cmd/buf/command/format/format.go

```diff
@@ -0,0 +1,388 @@
+				retErr = multierr.Append(retErr, moduleFile.Close())
+			}()
+			break
+		}
+		if moduleFile == nil {
+			// This will only happen if a buf.work.yaml exists in a parent
+			// directory, but it does not contain the target file.
+			//
+			// This is also a problem for other commands that interact
+			// with buffetch.ProtoFileRef.
+			//
+			// TODO: Fix the buffetch.ProtoFileRef so that it works in
+			// these situtations.
+			return fmt.Errorf("input %s was not found - is the directory containing this file defined in your buf.work.yaml?", container.Arg(0))
+		}
+		module, err = bufmodule.ModuleWithTargetPaths(
+			module,
+			[]string{
+				moduleFile.Path(),
+			},
+			nil, // Nothing to exclude.
+		)
+		if err != nil {
+			return err
+		}
+		return formatModule(
+			ctx,
+			container,
+			runner,
+			module,
+			readWriteBucket,
+			flags.ErrorFormat,
+			flags.Diff,
+			flags.Write,
+		)
+	}
+	for _, moduleConfig := range moduleConfigs {
+		if err := formatModule(
+			ctx,
+			container,
+			runner,
+			moduleConfig.Module(),
+			readWriteBucket,
+			flags.ErrorFormat,
+			flags.Diff,
+			flags.Write,
+		); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// formatModule formats the module's target files and writes them to the
+// writeBucket, if any. If diff is true, the diff between the original and
+// formatted files is written to stdout.
+func formatModule(
+	ctx context.Context,
+	container appflag.Container,
+	runner command.Runner,
+	module bufmodule.Module,
+	writeBucket storage.WriteBucket,
+	errorFormat string,
+	diff bool,
+	rewrite bool,
+) (retErr error) {
+	// Note that external paths are set properly for the files in this read bucket.
+	formattedReadBucket, err := bufformat.Format(ctx, module)
+	if err != nil {
+		return err
+	}
+	var originalReadWriteBucket storage.ReadWriteBucket
+	if diff || rewrite {
+		originalReadWriteBucket = storagemem.NewReadWriteBucket()
+		if err := bufmodule.TargetModuleFilesToBucket(
+			ctx,
+			module,
+			originalReadWriteBucket,
+		); err != nil {
+			return err
+		}
+	}
+	if diff {
+		if err := storage.Diff(
+			ctx,
+			runner,
+			container.Stdout(),
+			originalReadWriteBucket,
+			formattedReadBucket,
+			storage.DiffWithExternalPaths(), // No need to set prefixes as the buckets are from the same location.
+		); err != nil {
+			return err
+		}
+		if writeBucket == nil || !rewrite {
+			// If the user specified --diff and has not explicitly overridden
+			// the --output or rewritten the sources in-place with --write, we
+			// can stop here.
+			return nil
+		}
+	}
+	if rewrite {
+		// Rewrite the sources in place.
+		return storage.WalkReadObjects(
+			ctx,
+			originalReadWriteBucket,
+			"",
+			func(readObject storage.ReadObject) error {
+				formattedReadObject, err := formattedReadBucket.Get(ctx, readObject.Path())
+				if err != nil {
+					return err
+				}
+				file, err := os.OpenFile(readObject.ExternalPath(), os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0644)
```

Comment:
We have `storage.Copy` to a storage OS bucket, why can't we use that?

Reply:
In short, it's a lot simpler to use this approach since it can be applied to all of our output types in the same way. I left a comment to justify the current implementation - let me know what you think.